src/main/java/no/unit/nva/amazon/s3/ListPartsHandler.java

```diff
@@ -0,0 +1,140 @@
+import com.amazonaws.services.s3.model.ListPartsRequest;
+import com.amazonaws.services.s3.model.PartListing;
+import com.amazonaws.services.s3.model.PartSummary;
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static no.unit.nva.amazon.s3.GatewayResponse.ACCESS_CONTROL_ALLOW_ORIGIN;
+import static no.unit.nva.amazon.s3.GatewayResponse.BODY_KEY;
+import static org.apache.http.HttpHeaders.CONTENT_TYPE;
+import static org.apache.http.HttpStatus.SC_BAD_REQUEST;
+import static org.apache.http.HttpStatus.SC_INTERNAL_SERVER_ERROR;
+import static org.apache.http.HttpStatus.SC_OK;
+import static org.apache.http.entity.ContentType.APPLICATION_JSON;
+
+
+public class ListPartsHandler implements RequestHandler<Map<String, Object>, GatewayResponse> {
+
+    public static final String ALLOWED_ORIGIN_KEY = "ALLOWED_ORIGIN";
+    public static final String S3_UPLOAD_BUCKET_KEY = "S3_UPLOAD_BUCKET";
+
+    public final transient String bucketName;
+    private final transient String allowedOrigin;
+    private final transient AmazonS3 s3Client;
+
+    public ListPartsHandler() {
+        this(new Environment(), createAmazonS3Client());
+    }
+
+
+    /**
+     * Construct for lambda eventhandler to create an upload request for S3.
+     */
+    public ListPartsHandler(Environment environment, AmazonS3 s3Client) {
+        this.allowedOrigin = environment.get(ALLOWED_ORIGIN_KEY).orElseThrow(IllegalStateException::new);
+        this.bucketName = environment.get(S3_UPLOAD_BUCKET_KEY).orElseThrow(IllegalStateException::new);
+        this.s3Client = s3Client;
+    }
+
+    public static AmazonS3 createAmazonS3Client() {
+        return AmazonS3ClientBuilder.standard()
+                .build();
+    }
+
+    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
+    @Override
+    public GatewayResponse handleRequest(Map<String, Object> input, Context context) {
+
+        final GatewayResponse response = new GatewayResponse();
+        response.setHeaders(headers());
+
+
+        ListPartsRequestBody requestBody = null;
+        try {
+            requestBody = checkParameters(input);
+        } catch (JsonSyntaxException | ParameterMissingException e) {
+            response.setErrorBody(e.getMessage());
+            response.setStatusCode(SC_BAD_REQUEST);
+            System.out.println(response);
+            return response;
+        } catch (Exception e) {
+            response.setErrorBody(e.getMessage());
+            response.setStatusCode(SC_INTERNAL_SERVER_ERROR);
+            System.out.println(response);
+            return response;
+        }
+
+        try {
+
+            ListPartsRequest listPartsRequest = new ListPartsRequest(bucketName, requestBody.key, requestBody.uploadId);
+
+            PartListing partListing = s3Client.listParts(listPartsRequest);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

This code might not produce accurate results if the operation returns paginated results instead of all results. Consider adding another call to check for additional results.